### DISCUSSION TITLE
Merge different TUs

### DISCUSSION THREAD

**Discussion by anon767** [Fraunhofer-AISEC/cpg#660] [Ideas]
Hi,
I extracted the following question from: https://github.com/Fraunhofer-AISEC/cpg/discussions/655 because it was hardly related to the original post:

Is there a way to merge translationUnitDeclarations from different analysis passes? E.g. each file is analyzed and parsed separately and then merged into a single TranslationResult?
Then it would be possible to run a CallResolver pass over the merged translationDeclarations. When I tried that, I wasn't too sure how to populate the Scopes correctly.

Thanks :)

**Comment by oxisto** [comment]
Is this still an issue with the improved performance? You should be able to just analyse all your files in parallel mode now.

**Reply by anon767** [reply]
I thought it would be a cool feature if for example calculating the entire CPG for a large project is tedious. So one could calculate it once and then just recalculate changed TranslationUnits.
The idea reminds me of Datalog Static Analyzers like soufflé or [Laddder](https://dl.acm.org/doi/10.1145/3453483.3454026) .

As this is a discussion I'd love to hear what you think

**Reply by oxisto** [reply]
This is something that we have explored in a master thesis. https://www.sec.in.tum.de/i20/student-work/incremental-construction-of-code-property-graphs-1

The supplementing code has not been integrated into the main project (yet). In theory that is possible, however this is not something we are currently focusing on. Instead we are trying to significantly reduce the time it takes to calculate the CPG for large projects.

**Reply by anon767** [reply]
Alright good to know

**Reply by oxisto** [reply]
In that sense, if you have any large-scale projects that you can share or point us to, we are happy to have a look at potential performance issues / gains.

**Reply by anon767** [reply]
> The supplementing code has not been integrated into the **_main project_** (yet). 

just out of couriosity, is the code or the thesis online available?
